src/client/Inode.cc

```diff
@@ -219,9 +219,27 @@ void Inode::try_touch_cap(mds_rank_t mds)
   }
 }
 
-bool Inode::caps_issued_mask(unsigned mask)
+/**
+ * caps_issued_mask - check whether we have all of the caps in the mask
+ * @mask: mask to check against
+ * @allow_impl: whether the caller can also use caps that are implemented but not issued
+ *
+ * This is the bog standard "check whether we have the required caps" operation.
+ * Typically, we only check against the capset that is currently "issued".
+ * In other words, we ignore caps that have been revoked but not yet released.
+ *
+ * Some callers (particularly those doing attribute retrieval) can also make
+ * use of the full set of "implemented" caps to satisfy requests from the
+ * cache.
+ *
+ * Those callers should refrain from taking new references to implemented
```

Comment:
s/should/must/ ? :)

Reply:
Also: thanks for adding a good function doc!

Reply:
No problem! Anytime we start adding arguments like this, we need good docs.

I think I'd rather leave it as "should" for now, as I don't see a reason you absolutely couldn't take a reference to those caps. It's definitely not recommended, but I don't see how that would break anything. Am I missing something there?

Reply:
I don't know for certain, but I'm pretty sure by taking cap references after they've already been marked as revoked, we're essentially "leaking" them.
That is, the Client has already gone through the process of kicking anybody who's holding refs, so they should be dropping them. But if we take a ref here, nothing will kick at them. So the cap just gets held until this ref happens to get dropped in the future, which might be a loooong time.

Reply:
We generally only hold references when we're actively using a cap for something.

Taking a reference to hand out a new delegation after one of the caps has been revoked would almost certainly be problematic. However, taking an Fw reference to service a write in order to flush the cache so we can release Fb seems like it should be ok.